File: day6.php
<?php
$vnos = fopen("day6.txt", "r") or die("Unable to open file!");
$podatki = fread($vnos,filesize("day6.txt"));
fclose($vnos);
$razbito = preg_split("/\n/", $podatki);
$zacetek = 0;
for($i = 0; $i < count($razbito); $i++) {
	if(str_contains($razbito[$i], "^")) {
		for($k =0; $k < strlen($razbito[$i]); $k++) {
			if($razbito[$i][$k] == "^") {
				$zacetek = $i.' '.$k;
				$razbito[$i][$k] = "X";
				break 2;
			}
		}

	}
}
$zacetek = explode(" ", $zacetek);
$pot = "gor";
$isci = true;
$x = $zacetek[0];
$y = $zacetek[1];
while($isci) {
	if($pot == "gor") {
		if(isset($razbito[$x-1][$y]) && $razbito[$x-1][$y] !== "#") {
			$razbito[$x][$y] = "X";
			$x--;
		}else if(isset($razbito[$x-1][$y]) && $razbito[$x-1][$y] == "#") {
			$pot = "desno";
			$razbito[$x][$y] = "X";
			$y++;
		}else {
			$razbito[$x][$y] = "X";
			$isci = false;
		}
	}
	if($pot == "desno") {
		if(isset($razbito[$x][$y+1]) && $razbito[$x][$y+1] !== "#") {
			$razbito[$x][$y] = "X";
			$y++;
		} else if(isset($razbito[$x][$y+1]) && $razbito[$x][$y+1] == "#") {
			$razbito[$x][$y] = "X";
			$x++;
			$pot = "dol";
		}else {
			$razbito[$x][$y] = "X";
			$isci = false;
		}
	}
	if($pot == "dol") {
		if(isset($razbito[$x+1][$y]) && $razbito[$x+1][$y] !== "#") {
			$razbito[$x][$y] = "X";
			$x++;
		}else if(isset($razbito[$x+1][$y]) && $razbito[$x+1][$y] == "#") {
			$pot = "levo";
			$razbito[$x][$y] = "X";
			$y--;
		}else {
			$razbito[$x][$y] = "X";
			$isci = false;
		}
	}
	if($pot == "levo") {
		if(isset($razbito[$x][$y-1]) && $razbito[$x][$y-1] !== "#") {
			$razbito[$x][$y] = "X";
			$y--;
		} else if(isset($razbito[$x][$y-1]) && $razbito[$x][$y-1] == "#") {
			$razbito[$x][$y] = "X";
			$x--;
			$pot = "gor";
		}else {
			$razbito[$x][$y] = "X";
			$isci = false;
		}
	}
}
//part 1
$mapa = implode("\n", $razbito);
//map drawing
echo '<pre>';
echo $mapa;
echo '</pre>';
preg_match_all("/X/", $mapa, $izpis);
echo 'part1: '.count($izpis[0])-1;

//part 2 - still not optimal - will improve later
function ovire($mapa, $zX, $zY, $x, $y, $pot) {
	$pot = "gor";
	$mapa[$x][$y] = "#";
	$x = $zX;
	$y = $zY;
	$obiskane = array();
	$loop = false;
	while(true) {
		$koordinate = "{$x},{$y},{$pot}";
		if(array_search($koordinate, $obiskane)) {
			$izris = implode("\n", $mapa);
			$loop = true;
			break;
		}
		$obiskane[] = $koordinate;
		if($x -1 < 0 || $x + 1 == count($mapa) || $y - 1 < 0 || $y + 1 == count($mapa)) {
			break;
		}
		if($pot == "gor") {
			if($mapa[$x-1][$y] !== "#") {
				$x-= 1;
			}else if($mapa[$x-1][$y] == "#") {
				$pot = "desno";
				$y+= 1;
			}
		}
		if($pot == "desno") {
			if($mapa[$x][$y+1] !== "#") {
				$y+= 1;
			} else if($mapa[$x][$y+1] == "#") {
				$x+= 1;
				$pot = "dol";
			}
		}
		if($pot == "dol") {
			if($mapa[$x+1][$y] !== "#") {
				$x+=1;
			}else if($mapa[$x+1][$y] == "#") {
				$pot = "levo";
				$y-= 1;
			}
		}
		if($pot == "levo") {
			if($mapa[$x][$y-1] !== "#") {
				$y-= 1;
			} else if($mapa[$x][$y-1] == "#") {
				$x-= 1;
				$pot = "gor";
			}
		}
	}
	return $loop;
}
$ovir = 0;
$x = $zacetek[0];
$y = $zacetek[1];
foreach($path as $posamezna) {
	$cords = explode(",", $posamezna);
	$zX = $cords[0];
	$zY = $cords[1];
	$pot = $cords[2];
	if(ovire($zemljevid, $x, $y, $zX, $zY, $pot)) {
		$ovir++;
	}
}
echo '<br/>';
echo 'part2: '.$ovir-1;

?>

File: day6.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

namespace day6
{

    internal class Program
    {
        public static bool Ovire(string[] mapa, int y, int x, int zY, int zX)
        {
            /* zY in zX = # */
            string pot = "gor";
            var vrstica = mapa[zY];
            var nVrstica = "";           
            for(int i = 0; i < vrstica.Length; i++)
            {
                if(i == zX)
                {
                    nVrstica += '#';
                }else
                {
                    nVrstica += vrstica[i];
                }
            }
            mapa[zY] = nVrstica;
            List<string> obiskane = new List<string>();
            bool loop = false;
            while(true)
            {
                //Console.WriteLine($"{y}, {x} - {pot}");
                string koordinate = $"{y},{x},{pot}";
                if(obiskane.Contains(koordinate))
                {
                    mapa[zY] = vrstica;
                    loop = true;
                    break;
                }
                obiskane.Add(koordinate);
                if(x-1 < 0 || x+1 == mapa[y].Trim().Length || y-1 < 0 || y+1 == mapa.Length)
                {
                    mapa[zY] = vrstica;
                    break;
                }
                if(pot == "gor")
                {
                    if (mapa[y - 1][x] != '#')
                    {
                        y--;
                    }else if (mapa[y - 1][x] == '#')
                    {
                        pot = "desno";
                        x++;
                    }
                }
                if(pot == "desno")
                {
                    if ( mapa[y][x+1] != '#')
                    {
                        x++;
                    }else if ( mapa[y][x+1] == '#')
                    {
                        pot = "dol";
                        y++;
                    }
                }
                if(pot == "dol")
                {
                    if (mapa[y + 1][x] != '#')
                    {
                        y++;
                    }else if (mapa[y+1][x] == '#')
                    {
                        pot = "levo";
                        x--;
                    }
                }
                if(pot == "levo")
                {
                    if ( mapa[y][x-1] != '#')
                    {
                        x--;
                    }else if ( mapa[y][x-1] == '#')
                    {
                        pot = "gor";
                        y--;
                    }
                }
            }
            mapa[zY] = vrstica;
            return loop;
        }

        static void Main(string[] args)
        {
            string input = @"day6.txt";
            if (File.Exists(input))
            {
                string readInput = File.ReadAllText(input);
                string[] vrstica = Regex.Split(readInput.Trim(), "\n");
                int y = 0; 
                int zY = 0;
                int x = 0; 
                int zX = 0;
                for (int i = 0; i < vrstica.Length; i++)
                {
                    for (int j = 0; j < vrstica[i].Length; j++)
                    {
                        if (vrstica[i][j] == '^')
                        {
                            y = i;
                            x = j;
                            zY = i;
                            zX = j;
                            break;
                        }
                    }
                }
                List<string> hoja = new List<string>();
                List<string> mapa = new List<string>();
                var pot = "gor";
                bool isci = true;
                while (isci)
                {
                    mapa.Add(y +","+ x +","+ pot);
                    if (pot == "gor")
                    {
                        if (y - 1 >= 0 && vrstica[y-1][x] != '#')
                        {
                            if (!hoja.Contains(y + " " + x))
                            {
                                hoja.Add(y + " " + x);
                            }
                            y--;
                        }else if (y - 1 >= 0 && vrstica[y-1][x] == '#')
                        {
                            pot = "desno";
                            if (!hoja.Contains(y + " " + x))
                            {
                                hoja.Add(y + " " + x);
                            }
                            x++;
                        }else
                        {
                            if (!hoja.Contains(y + " " + x))
                            {
                                hoja.Add(y + " " + x);
                            }
                            isci = false;
                        }

                    }
                    if(pot == "desno")
                    {
                        if (x + 1 < vrstica[y].Length && vrstica[y][x + 1] != '#')
                        {
                            if (!hoja.Contains(y + " " + x))
                            {
                                hoja.Add(y + " " + x);
                            }
                            x++;
                        }else if (x + 1 < vrstica[y].Length && vrstica[y][x+1] == '#')
                        {
                            pot = "dol";
                            if (!hoja.Contains(y + " " + x))
                            {
                                hoja.Add(y + " " + x);
                            }
                            y++;
                        }else
                        {
                            if (!hoja.Contains(y + " " + x))
                            {
                                hoja.Add(y + " " + x);
                            }
                            isci = false;
                        }
                    }
                    if(pot == "dol")
                    {
                        if (y + 1 < vrstica.Length && vrstica[y + 1][x] != '#')
                        {
                            if (!hoja.Contains(y + " " + x))
                            {
                                hoja.Add(y + " " + x);
                            }
                            y++;
                        }else if (y + 1 < vrstica.Length && vrstica[y + 1][x] == '#')
                        {
                            pot = "levo";
                            if (!hoja.Contains(y + " " + x))
                            {
                                hoja.Add(y + " " + x);
                            }
                            x--;
                        }else
                        {
                            if (!hoja.Contains(y + " " + x))
                            {
                                hoja.Add(y + " " + x);
                            }
                            isci = false;
                        }
                    }
                    if(pot == "levo")
                    {
                        if (x - 1 >= 0 && vrstica[y][x - 1] != '#')
                        {
                            if (!hoja.Contains(y + " " + x))
                            {
                                hoja.Add(y + " " + x);
                            }
                            x--;
                        }else if (x - 1 >= 0 && vrstica[y][x-1] == '#')
                        {
                            pot = "gor";
                            if (!hoja.Contains(y + " " + x))
                            {
                                hoja.Add(y + " " + x);
                            }
                            y--;
                        }else
                        {
                            if (!hoja.Contains(y + " " + x))
                            {
                                hoja.Add(y + " " + x);
                            }
                            isci = false;
                        }
                    }
                }
                Console.WriteLine("part 1:" + hoja.Count);
                
                //part 2
                y = zY;
                x = zX;
                int ovire = 0;
                foreach (var m in hoja)
                {
                    //Console.WriteLine(m);
                    var koordinate = m.Split(' ');
                    zY = Int32.Parse(koordinate[0]);
                    zX = Int32.Parse(koordinate[1]);
                    //pot = koordinate[2];
                    if (zX == x && zY == y || vrstica[zY][zX] == '#')
                    {
                        continue;
                    }
                    if (Ovire(vrstica, y, x, zY, zX))
                    {
                        ovire++;
                    }
                }
                Console.WriteLine("part 2:" + ovire);
            }
        }
    }
}
